pkg/wwan/mmagent/agent.go

```diff
@@ -0,0 +1,1451 @@
+			a.publishWwanStatus()
+
+		case <-publishMetricsTimer.C:
+			a.publishMetrics()
+
+		case <-stillRunning.C:
+			if time.Since(a.mmClient.LastSeenMM()) >= wdTouchPeriod {
+				if _, err := a.mmClient.GetMMVersion(); err != nil {
+					a.log.Warnf("Failed to get MM version (process crashed?): %v", err)
+				}
+			}
+		}
+
+		// Here we implement watchdog detection for both this agent and the ModemManager.
+		if time.Since(a.mmClient.LastSeenMM()) < wdTouchPeriod {
+			a.ps.StillRunning(agentName, warningTime, errorTime)
+		}
+	}
+}
+
+func (a *MMAgent) ignoreNonGlobalKey(key string) bool {
+	if key != "global" {
+		a.log.Warnf("Ignoring pubsub message key=%s", key)
+		return true
+	}
+	return false
+}
+
+func (a *MMAgent) handleGlobalConfigCreate(_ interface{}, key string, arg interface{}) {
+	if a.ignoreNonGlobalKey(key) {
+		return
+	}
+	a.applyGlobalConfig(arg.(types.ConfigItemValueMap))
+}
+
+func (a *MMAgent) handleGlobalConfigModify(_ interface{}, key string, arg, _ interface{}) {
+	if a.ignoreNonGlobalKey(key) {
+		return
+	}
+	a.applyGlobalConfig(arg.(types.ConfigItemValueMap))
+}
+
+func (a *MMAgent) handleGlobalConfigDelete(_ interface{}, key string, arg interface{}) {
+	if a.ignoreNonGlobalKey(key) {
+		return
+	}
+	a.applyGlobalConfig(*types.DefaultConfigItemValueMap())
+}
+
+func (a *MMAgent) handleWwanConfigCreate(_ interface{}, key string, arg interface{}) {
+	if a.ignoreNonGlobalKey(key) {
+		return
+	}
+	a.applyWwanConfig(arg.(types.WwanConfig))
+}
+
+func (a *MMAgent) handleWwanConfigModify(_ interface{}, key string, arg, _ interface{}) {
+	if a.ignoreNonGlobalKey(key) {
+		return
+	}
+	a.applyWwanConfig(arg.(types.WwanConfig))
+}
+
+func (a *MMAgent) handleWwanConfigDelete(_ interface{}, key string, _ interface{}) {
+	if a.ignoreNonGlobalKey(key) {
+		return
+	}
+	a.applyWwanConfig(types.WwanConfig{})
+}
+
+func (a *MMAgent) applyGlobalConfig(config types.ConfigItemValueMap) {
+	a.globalConfig = config
+	prevLogLevel := a.logger.GetLevel()
+	agentlog.HandleGlobalConfig(a.log, a.subGlobalConfig, agentName,
+		a.CLIParams().DebugOverride, a.logger)
+	if a.logger.GetLevel() != prevLogLevel || !a.gcInitialized {
+		err := a.mmClient.SetMMLogLevel(a.logger.GetLevel())
+		if err == nil {
+			a.log.Noticef("Changed ModemManager log level to %v", a.logger.GetLevel())
+		} else {
+			a.log.Warnf("Failed to set ModemManager log level to %v: %v",
+				a.logger.GetLevel(), err)
+		}
+	}
+	// Publish location info 2x more often (at most) than zedagent publishes
+	// to applications and controller.
+	locPublishCloudPeriod := time.Second *
+		time.Duration(a.globalConfig.GlobalValueInt(types.LocationCloudInterval))
+	locPublishAppPeriod := time.Second *
+		time.Duration(a.globalConfig.GlobalValueInt(types.LocationAppInterval))
+	publishInterval := locPublishAppPeriod
+	if locPublishCloudPeriod < publishInterval {
+		// This is quite unlikely config.
+		publishInterval = locPublishCloudPeriod
+	}
+	publishInterval = publishInterval >> 1
```

Comment:
Why do we square publishInterval here?

Reply:
This is division by 2.
See reasoning here: https://github.com/milan-zededa/eve/blob/modem-manager/pkg/wwan/mmagent/agent.go#L504-L505